sql/core/src/main/scala/org/apache/spark/sql/execution/python/VectorizedPythonRunner.scala

```diff
@@ -0,0 +1,329 @@
+  private val pythonExec = funcs.head.funcs.head.pythonExec
+  private val pythonVer = funcs.head.funcs.head.pythonVer
+
+  // TODO: support accumulator in multiple UDF
+  private val accumulator = funcs.head.funcs.head.accumulator
+
+  // todo: return column batch?
+  def compute(
+      inputRows: Iterator[InternalRow],
+      schema: StructType,
+      partitionIndex: Int,
+      context: TaskContext): Iterator[InternalRow] = {
+    val startTime = System.currentTimeMillis
+    val env = SparkEnv.get
+    val localdir = env.blockManager.diskBlockManager.localDirs.map(f => f.getPath()).mkString(",")
+    envVars.put("SPARK_LOCAL_DIRS", localdir) // it's also used in monitor thread
+    if (reuse_worker) {
+      envVars.put("SPARK_REUSE_WORKER", "1")
+    }
+    val worker: Socket = env.createPythonWorker(pythonExec, envVars.asScala.toMap)
+    // Whether is the worker released into idle pool
+    @volatile var released = false
+
+    // Start a thread to feed the process input from our parent's iterator
+    val writerThread = new WriterThread(
+      env, worker, inputRows, schema, partitionIndex, context)
+
+    context.addTaskCompletionListener { context =>
+      writerThread.shutdownOnTaskCompletion()
+      if (!reuse_worker || !released) {
+        try {
+          worker.close()
+        } catch {
+          case e: Exception =>
+            logWarning("Failed to close worker socket", e)
+        }
+      }
+    }
+
+    writerThread.start()
+
+    val stream = new DataInputStream(new BufferedInputStream(worker.getInputStream, bufferSize))
+
+    val allocator = ArrowUtils.rootAllocator.newChildAllocator(
+      s"stdin reader for $pythonExec", 0, Long.MaxValue)
+    val reader = new ArrowStreamReader(stream, allocator)
+
+    new Iterator[InternalRow] {
+      private val root = reader.getVectorSchemaRoot
+      private val vectors = root.getFieldVectors.asScala.map { vector =>
+        new ArrowColumnVector(vector)
+      }.toArray[ColumnVector]
+
+      var closed = false
+
+      context.addTaskCompletionListener { _ =>
+        // todo: we need something like `read.end()`, which release all the resources, but leave
+        // the input stream open. `reader.close` will close the socket and we can't reuse worker.
+        // So here we simply not close the reader, which is problematic.
+        if (!closed) {
+          root.close()
+          allocator.close()
+        }
+      }
+
+      private[this] var batchLoaded = true
+      private[this] var currentIter: Iterator[InternalRow] = Iterator.empty
+
+      override def hasNext: Boolean = batchLoaded && (currentIter.hasNext || loadNextBatch()) || {
+        root.close()
+        allocator.close()
+        closed = true
+        false
+      }
+
+      private def loadNextBatch(): Boolean = {
+        batchLoaded = reader.loadNextBatch()
+        if (batchLoaded) {
+          val batch = new ColumnarBatch(schema, vectors, root.getRowCount)
```

Comment:
A side note: How is performance of ColumnarBatch in terms of converting to a `Iterator[InternalRow]`? As far as I remember it doesn't return unsafe row at the moment, right?

Reply:
That's right, it doesn't return unsafe row.
But I believe it's performant enough because it can return values in row directly from column vectors without copying to unsafe row.

Reply:
Interesting, does this mean it's more performant than copying the column vectors into unsafe row? Because to get any value out, it has to access the memory region of column vectors any way, therefore copying bytes from column vectors into unsafe row don't improve performance?

Reply:
Why copying bytes from column vectors into unsafe row can improve performance? Isn't direct access the data from the column vectors faster without the cost of copying bytes?